### PR TITLE
gh-117142: ctypes: Unify meta tp slot functions

### PR DESCRIPTION
**Pull request by neonene**
Integrates the following ctypes meta tp slot functions:
* `CDataType_traverse()` into `CType_Type_traverse()`.
* `CDataType_clear()` into `CType_Type_clear()`.
* `CDataType_dealloc()` into `CType_Type_dealloc()`.
* `CDataType_repeat()` into `CType_Type_repeat()`.

<br>
<details>
<summary>Previous description (expand):</summary>

>An attempt to introduce multi-phase initialization:
>
>* Keep using a global state.
>
>* Integrate `CDataType_*` functions into `CType_Type_*()`.
>
>* Make `module_traverse()` visit almost all global variables to invoke `module_clear()`, >leaving aside `malloc_closure.c (free_list)` for now.  The `free_list` is related to >`CThunkObject_dealloc()` and I'm not sure if it also needs isolation.
</details>
<br>

* Issue: gh-117142

### CODE REVIEW

File: Modules/_ctypes/_ctypes.c [hunk 1]
```diff
@@ -1296,34 +1262,29 @@ PyCPointerType_from_param(PyObject *type, PyObject *value)
...
 
 static PyMethodDef PyCPointerType_methods[] = {
-    { "from_address", CDataType_from_address, METH_O, from_address_doc },
-    { "from_buffer", CDataType_from_buffer, METH_VARARGS, from_buffer_doc, },
-    { "from_buffer_copy", CDataType_from_buffer_copy, METH_VARARGS, from_buffer_copy_doc, },
-    { "in_dll", CDataType_in_dll, METH_VARARGS, in_dll_doc},
+    { "from_address", CTypeType_from_address, METH_O, from_address_doc },
+    { "from_buffer", CTypeType_from_buffer, METH_VARARGS, from_buffer_doc, },
+    { "from_buffer_copy", CTypeType_from_buffer_copy, METH_VARARGS, from_buffer_copy_doc, },
+    { "in_dll", CTypeType_in_dll, METH_VARARGS, in_dll_doc},
```
Comment (encukou) [inline Modules/_ctypes/_ctypes.c]: Shouldn't these should be inherited too?

File: Modules/_ctypes/_ctypes.c [hunk 2]
```diff
@@ -2278,20 +2270,17 @@ PyCSimpleType_from_param(PyObject *type, PyObject *value)
...
 static PyMethodDef PyCSimpleType_methods[] = {
     { "from_param", PyCSimpleType_from_param, METH_O, from_param_doc },
-    { "from_address", CDataType_from_address, METH_O, from_address_doc },
-    { "from_buffer", CDataType_from_buffer, METH_VARARGS, from_buffer_doc, },
-    { "from_buffer_copy", CDataType_from_buffer_copy, METH_VARARGS, from_buffer_copy_doc, },
-    { "in_dll", CDataType_in_dll, METH_VARARGS, in_dll_doc},
+    { "from_address", CTypeType_from_address, METH_O, from_address_doc },
+    { "from_buffer", CTypeType_from_buffer, METH_VARARGS, from_buffer_doc, },
+    { "from_buffer_copy", CTypeType_from_buffer_copy, METH_VARARGS, from_buffer_copy_doc, },
+    { "in_dll", CTypeType_in_dll, METH_VARARGS, in_dll_doc},
```
Comment (encukou) [inline Modules/_ctypes/_ctypes.c]: And these

File: Modules/_ctypes/_ctypes.c [hunk 3]
```diff
@@ -2544,11 +2533,8 @@ PyCFuncPtrType_init(PyObject *self, PyObject *args, PyObject *kwds)
 
 static PyType_Slot pycfuncptr_type_slots[] = {
     {Py_tp_doc, PyDoc_STR("metatype for C function pointers")},
-    {Py_tp_methods, CDataType_methods},
+    {Py_tp_methods, CTypeType_methods},
```
Comment (encukou) [inline Modules/_ctypes/_ctypes.c]: And this

File: Modules/_ctypes/_ctypes.c [hunk 4]
```diff
@@ -1032,7 +1028,7 @@ UnionType_setattro(PyObject *self, PyObject *key, PyObject *value)
 static PyType_Slot pycstruct_type_slots[] = {
     {Py_tp_setattro, PyCStructType_setattro},
     {Py_tp_doc, PyDoc_STR("metatype for the CData Objects")},
-    {Py_tp_methods, CTypeType_methods},
+    {Py_tp_methods, ctype_methods},
```
Comment (encukou) [inline Modules/_ctypes/_ctypes.c]: And all of the `ctype_methods`

**Review by encukou — COMMENTED** [review]
This looks great, thank you! Just hold off on the `PyModuleDef` changes while the state is global -- see below.

---

Consider splitting the inheritance to a separate PR -- it's not only easier to review, but if something goes wrong we can revert it separately.

---

This still uses the global state, so switching to multi-phase init won't work right. For example:
```
$ ./python
Python 3.13.0a5+ (heads/ctypes_multi:6498c94f686, Mar 22 2024, 15:48:20) [GCC 13.2.1 20240316 (Red Hat 13.2.1-7)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ctypes
>>> import sys
>>> del sys.modules['_ctypes']
>>> import _ctypes
>>> exit()
Python/gc.c:93: gc_decref: Assertion "gc_get_refs(g) > 0" failed: refcount is too small
[...]
Aborted (core dumped)
```

There are ways around this, but IMO the best one is to fully isolate module state before switching to multi-phase init.

**Comment by neonene** [comment]
Thank you for reviewing and pointing that out. I would never notice that error.
Before reconsidering this, would you mind taking a look at my full isolation try: #117181?

**Comment by neonene** [comment]
See also #117189.

**Comment by encukou** [comment]
OK, I'll look at #117189 first (some time this week).

**Comment by neonene** [comment]
I think this is ready for review, and I can separate or revert any commits.

**Comment by neonene** [comment]
#117181 (multi-phase init) is ready as well.

**Comment by encukou** [comment]
Thank you!

It looks like the `CDataType_methods` could also be moved to the superclass, and inherited. But this PR is an improvement as is :)

**Pull request merged**